src/tallies/tally_scoring.cpp

```diff
@@ -0,0 +1,2347 @@
+      // touch any of the remaining filters.
+      ++match.i_bin_;
+      done_looping = false;
+      break;
+    } else {
+      // This bin cannot be incremented so reset it and continue to the next
+      // filter.
+      match.i_bin_ = 0;
+    }
+  }
+
+  if (done_looping) {
+    // We have visited every valid combination.  All done!
+    index_ = -1;
+  } else {
+    // The loop found a new valid combination.  Compute the corresponding
+    // index and weight.
+    compute_index_weight();
+  }
+
+  return *this;
+}
+
+void
+FilterBinIter::compute_index_weight()
+{
+  index_ = 0;
+  weight_ = 1.;
+  for (auto i = 0; i < tally_.filters().size(); ++i) {
+    auto i_filt = tally_.filters(i);
+    auto& match {simulation::filter_matches[i_filt]};
+    auto i_bin = match.i_bin_;
+    index_ += match.bins_[i_bin] * tally_.strides(i);
+    weight_ *= match.weights_[i_bin];
+  }
+}
+
+//==============================================================================
+// Non-member functions
+//==============================================================================
+
+//! Helper function used to increment tallies with a delayed group filter.
+
+void
+score_fission_delayed_dg(int i_tally, int d_bin, double score, int score_index)
+{
+  // Save the original delayed group bin
+  auto& tally {*model::tallies[i_tally]};
+  auto i_filt = tally.filters(tally.delayedgroup_filter_);
+  auto& dg_match {simulation::filter_matches[i_filt]};
+  auto i_bin = dg_match.i_bin_;
+  auto original_bin = dg_match.bins_[i_bin];
+  dg_match.bins_[i_bin] = d_bin;
+
+  // Determine the filter scoring index
+  auto filter_index = 0;
+  for (auto i = 0; i < tally.filters().size(); ++i) {
+    auto i_filt = tally.filters(i);
+    auto& match {simulation::filter_matches[i_filt]};
+    auto i_bin = match.i_bin_;
+    filter_index += match.bins_[i_bin] * tally.strides(i);
+  }
+
+  // Update the tally result
+  #pragma omp atomic
+  tally.results_(filter_index, score_index, RESULT_VALUE) += score;
+
+  // Reset the original delayed group bin
+  dg_match.bins_[i_bin] = original_bin;
+}
+
+//! Helper function to retrieve fission q value from a nuclide
+
+double get_nuc_fission_q(const Nuclide& nuc, const Particle* p, int score_bin)
+{
+  if (score_bin == SCORE_FISS_Q_PROMPT) {
+    if (nuc.fission_q_prompt_) {
+      return (*nuc.fission_q_prompt_)(p->E_last_);
+    }
+  } else if (score_bin == SCORE_FISS_Q_RECOV) {
+    if (nuc.fission_q_recov_) {
+      return (*nuc.fission_q_recov_)(p->E_last_);
+    }
+  }
+  return 0.0;
+}
+
+//! Helper function to score fission energy
+//
+//! Pulled out to support both the fission_q scores and energy deposition
+//! score
+
+double score_fission_q(const Particle* p, int score_bin, const Tally& tally,
+  double flux, int i_nuclide, double atom_density)
+{
+  if (tally.estimator_ == ESTIMATOR_ANALOG) {
+    const Nuclide& nuc {*data::nuclides[p->event_nuclide_]};
```

Comment:
Thank you for dereferencing this, and turning it into a reference const variable.  Makes it significantly more readable.